addon/webextension/background/selectorLoader.js

```diff
@@ -65,12 +65,12 @@ this.selectorLoader = (function() {
   let loadingTabs = new Set();
 
   exports.loadModules = function(tabId, hasSeenOnboarding) {
-    let promise;
     loadingTabs.add(tabId);
+    let promise = incognitoCheck(tabId);
     if (hasSeenOnboarding) {
-      promise = executeModules(tabId, standardScripts.concat(selectorScripts));
+      promise = promise.then(executeModules(tabId, standardScripts.concat(selectorScripts)));
     } else {
-      promise = executeModules(tabId, standardScripts.concat(onboardingScripts).concat(selectorScripts));
+      promise = promise.then(executeModules(tabId, standardScripts.concat(onboardingScripts).concat(selectorScripts)));
```

Comment:
This isn't right, is it? That is, you want `promise` (aka `incognitoCheck`) to complete, *then* call `executeModules` – but this code doesn't do that. It does `incognitoCheckPromise.then(executeModulesPromise)` instead of `incognitoCheckPromise.then(callbackThatRunsExecuteModules)`

Reply:
I *think* it's correct as written, since `executeModules` returns a promise. Seems like the code works. Am I missing something?

Reply:
I don't know what the behavior of `.then()` is supposed to be when it gets a promise, but at best this is equivalent to:

```
promise = Promise.all([promise, executeModules(...)])
```

That is, as soon as the promise is created it's going to "start". Probably the promises will run in the right order because they have to run in some order, and order of creation is mostly likely, but it's undefined behavior. If you want executeModules to run *after* the other promise, then you have to only create the executeModules promise after the first promise completes.

I think the behavior will be clearer if you put an error into the incognito code – which should (but currently I believe won't) stop the rest from running.

Reply:
ah, I see what you mean. `.then` expects a function which returns a promise when invoked, not the promise itself